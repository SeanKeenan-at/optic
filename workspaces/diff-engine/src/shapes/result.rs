use crate::shapes::{JsonTrail, ShapeTrail};
use serde::{Deserialize, Serialize};
use std::collections::hash_map::DefaultHasher;
use std::hash::{Hash, Hasher};

#[derive(Debug, Deserialize, Serialize, Hash)]
pub enum ShapeDiffResult {
  #[serde(rename_all = "camelCase")]
  UnspecifiedShape {
    json_trail: JsonTrail,
    shape_trail: ShapeTrail,
  },
  #[serde(rename_all = "camelCase")]
  UnmatchedShape {
    json_trail: JsonTrail,
    shape_trail: ShapeTrail,
  },
}

impl ShapeDiffResult {
  pub fn fingerprint(&self) -> String {
    let mut hash_state = DefaultHasher::new();
    Hash::hash(self, &mut hash_state);
    format!("{:x}", hash_state.finish())
  }
}
