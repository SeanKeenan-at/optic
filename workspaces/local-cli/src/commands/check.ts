import { Command } from '@oclif/command';
import { verifyTask } from '../shared/verify/verify';

export default class Check extends Command {
  static description =
    'verify that Optic can run your tasks and monitor traffic';

  static args = [
    {
      name: 'taskName',
    },
  ];

  async run() {
    const { args } = this.parse(Check);
    const { taskName } = args;
    await verifyTask(this, taskName);
    process.exit(0);
  }
}
