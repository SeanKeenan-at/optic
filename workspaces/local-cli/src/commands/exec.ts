import { Command, flags } from '@oclif/command';
import {
  LocalCliTaskFlags,
  LocalTaskSessionWrapper,
  runCommandFlags,
} from '../shared/local-cli-task-runner';
import { cleanupAndExit, loadPathsAndConfig } from '@useoptic/cli-shared';
import { isCommandOnlyTask } from '@useoptic/cli-config';

export default class Exec extends Command {
  static description =
    'run a command and receives traffic over an ingest-service';

  static flags = runCommandFlags;

  static args = [
    {
      name: 'command',
    },
  ];

  async run() {
    const { args } = this.parse(Exec);
    const { flags } = this.parse(Exec);
    const { taskName } = args;
  }
}

export function runExec(
  cli: Command,
  command: string,
  flags: LocalCliTaskFlags
) {}
