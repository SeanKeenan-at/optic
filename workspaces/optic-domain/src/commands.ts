/*
TODO Replace with Rust constructors eventually
*/

// TODO fix any types in this file
export type CQRSCommand =
  | AddShapeType
  | AddQueryParametersType
  | AddRequestType
  | FieldFromShapeType
  | AddFieldType
  | ShapeProviderType
  | ProviderInShapeType
  | SetParameterShapeType
  | SetFieldShapeType
  | AddShapeParameterType
  | RemoveFieldType
  | AddPathParameterType
  | AddPathComponentType
  | AddContributionType
  | ShapedBodyDescriptorType
  | SetQueryParameteresShapeType
  | SetResponseBodyShapeType
  | SetRequestBodyShapeType
  | AddResponseByPathAndMethodType
  | PrunePathComponentsType
  | QueryParametersShapeDescriptor;

export type AddShapeType = ReturnType<typeof AddShape>;
export type AddQueryParametersType = ReturnType<typeof AddQueryParameters>;
export type AddRequestType = ReturnType<typeof AddRequest>;
export type FieldFromShapeType = ReturnType<typeof FieldShapeFromShape>;
export type AddFieldType = ReturnType<typeof AddField>;
export type ShapeProviderType = ReturnType<typeof ShapeProvider>;
export type ProviderInShapeType = ReturnType<typeof ProviderInShape>;
export type SetParameterShapeType = ReturnType<typeof SetParameterShape>;
export type SetFieldShapeType = ReturnType<typeof SetFieldShape>;
export type AddShapeParameterType = ReturnType<typeof AddShapeParameter>;
export type RemoveFieldType = ReturnType<typeof RemoveField>;
export type AddPathParameterType = ReturnType<typeof AddPathParameter>;
export type AddPathComponentType = ReturnType<typeof AddPathComponent>;
export type AddContributionType = ReturnType<typeof AddContribution>;
export type ShapedBodyDescriptorType = ReturnType<typeof ShapedBodyDescriptor>;
export type SetQueryParameteresShapeType = ReturnType<
  typeof SetQueryParametersShape
>;
export type QueryParametersShapeDescriptor = ReturnType<
  typeof QueryParametersShapeDescriptor
>;
export type SetResponseBodyShapeType = ReturnType<typeof SetResponseBodyShape>;
export type SetRequestBodyShapeType = ReturnType<typeof SetRequestBodyShape>;
export type AddResponseByPathAndMethodType = ReturnType<
  typeof AddResponseByPathAndMethod
>;
export type PrunePathComponentsType = ReturnType<typeof PrunePathComponents>;

export function AddShape(
  shapeId: string,
  baseShapeId: string,
  name: string = ''
) {
  return { AddShape: { shapeId, baseShapeId, name } };
}

export function AddQueryParameters(
  httpMethod: string,
  pathId: string,
  queryParametersId: string
) {
  return { AddQueryParameters: { httpMethod, pathId, queryParametersId } };
}

export function AddRequest(
  httpMethod: string,
  pathId: string,
  requestId: string
) {
  return { AddRequest: { httpMethod, pathId, requestId } };
}

export function FieldShapeFromShape(fieldId: string, shapeId: string) {
  return { FieldShapeFromShape: { fieldId, shapeId } };
}

export function AddField(
  fieldId: string,
  shapeId: string,
  name: string,
  shapeDescriptor: FieldFromShapeType
) {
  return { AddField: { fieldId, shapeId, name, shapeDescriptor } };
}

export function ShapeProvider(shapeId: string) {
  return { ShapeProvider: { shapeId } };
}

export function ProviderInShape(
  shapeId: string,
  providerDescriptor: any,
  consumingParameterId: string
) {
  return {
    ProviderInShape: { shapeId, providerDescriptor, consumingParameterId },
  };
}

export function SetParameterShape(shapeDescriptor: any) {
  return { SetParameterShape: { shapeDescriptor } };
}
export function SetFieldShape(shapeDescriptor: FieldFromShapeType) {
  return { SetFieldShape: { shapeDescriptor } };
}

export function AddShapeParameter(
  shapeParameterId: string,
  shapeId: string,
  name: string
) {
  return { AddShapeParameter: { shapeParameterId, shapeId, name } };
}

export function RemoveField(fieldId: string) {
  return { RemoveField: { fieldId } };
}

export function AddPathParameter(
  pathId: string,
  parentPathId: string,
  name: string
) {
  return {
    AddPathParameter: {
      pathId,
      parentPathId,
      name,
    },
  };
}

export function AddPathComponent(
  pathId: string,
  parentPathId: string,
  name: string
) {
  return {
    AddPathComponent: {
      pathId,
      parentPathId,
      name,
    },
  };
}

export function AddContribution(id: string, key: string, value: string) {
  return {
    AddContribution: {
      id,
      key,
      value,
    },
  };
}

export function ShapedBodyDescriptor(
  httpContentType: string,
  shapeId: string,
  isRemoved: boolean = false
) {
  return {
    httpContentType,
    shapeId,
    isRemoved,
  };
}

export function QueryParametersShapeDescriptor(
  shapeId: string,
  isRemoved: boolean = false
) {
  return {
    shapeId,
    isRemoved,
  };
}

export function SetQueryParametersShape(
  queryParametersId: string,
  shapeDescriptor: ReturnType<typeof QueryParametersShapeDescriptor>
) {
  return {
    SetQueryParametersShape: {
      queryParametersId,
      shapeDescriptor,
    },
  };
}

export function SetResponseBodyShape(
  responseId: string,
  bodyDescriptor: ReturnType<typeof ShapedBodyDescriptor>
) {
  return {
    SetResponseBodyShape: {
      responseId,
      bodyDescriptor,
    },
  };
}
export function SetRequestBodyShape(
  requestId: string,
  bodyDescriptor: ReturnType<typeof ShapedBodyDescriptor>
) {
  return {
    SetRequestBodyShape: {
      requestId,
      bodyDescriptor,
    },
  };
}

export function AddResponseByPathAndMethod(
  responseId: string,
  pathId: string,
  httpMethod: string,
  httpStatusCode: number
) {
  return {
    AddResponseByPathAndMethod: {
      responseId,
      pathId,
      httpMethod,
      httpStatusCode,
    },
  };
}

export function PrunePathComponents() {
  return {
    PrunePathComponents: {},
  };
}
