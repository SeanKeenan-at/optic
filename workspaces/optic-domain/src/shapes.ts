// Shape events
export enum ICoreShapeKinds {
  ObjectKind = '$object',
  ListKind = '$list',
  MapKind = '$map',
  OneOfKind = '$oneOf',
  AnyKind = '$any',
  StringKind = '$string',
  NumberKind = '$number',
  BooleanKind = '$boolean',
  NullableKind = '$nullable',
  OptionalKind = '$optional',
  UnknownKind = '$unknown',
}

export enum ICoreShapeInnerParameterNames {
  ListInner = '$listItem',
  NullableInner = '$nullableInner',
  OptionalInner = '$optionalInner',
}

// Optic engine shape types
export enum JsonLike {
  OBJECT = 'Object',
  ARRAY = 'Array',
  NULL = 'Null',
  STRING = 'String',
  NUMBER = 'Number',
  BOOLEAN = 'Boolean',
  UNDEFINED = 'Undefined',
}
