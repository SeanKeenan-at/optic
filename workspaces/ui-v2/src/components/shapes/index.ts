export * from './DepthContext';
export * from './OneOfTabs';
export * from './ShapePrimitive';
export * from './ShapeRenderContext';
export * from './ShapeRenderInterfaces';
export * from './ShapeRowBase';
export * from './SharedStyles';
