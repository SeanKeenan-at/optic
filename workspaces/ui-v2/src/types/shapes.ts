import { ChangeType } from './changes';

export interface IFieldRenderer {
  fieldId: string;
  name: string;
  shapeId: string;
  shapeChoices: IShapeRenderer[];
  required: boolean;
  changes: ChangeType | null;
  contributions: Record<string, string>;
  // TODO - move this into a different typing or figure out where this should live
  additionalAttributes?: string[];
}

export type IShapeRenderer =
  | {
      shapeId: string;
      jsonType: JsonLike.OBJECT;
      asArray?: undefined;
      asObject: {
        fields: IFieldRenderer[];
      };
    }
  | {
      shapeId: string;
      jsonType: JsonLike.ARRAY;
      asArray: IArrayRender;
      asObject?: undefined;
    }
  | {
      shapeId: string;
      jsonType: Exclude<JsonLike, JsonLike.OBJECT | JsonLike.ARRAY>;
      asArray?: undefined;
      asObject?: undefined;
    };

export interface IArrayRender {
  shapeChoices: IShapeRenderer[];
  shapeId: string;
}

// TODO remove this
export interface IObjectRender {
  fields: IFieldRenderer[];
}

export enum JsonLike {
  OBJECT = 'Object',
  ARRAY = 'Array',
  NULL = 'Null',
  STRING = 'String',
  NUMBER = 'Number',
  BOOLEAN = 'Boolean',
  UNDEFINED = 'Undefined',
}
