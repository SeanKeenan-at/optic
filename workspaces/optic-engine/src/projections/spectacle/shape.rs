use crate::projections::shape::{Edge, Node, ShapeProjection};
use crate::queries::spectacle::spec_choices::{ShapeChoice, ShapeChoiceQueries};
use crate::shapes::ShapeTrail;
use crate::state::shape::ShapeId;
use serde::Serialize;
use std::collections::BTreeMap;
use std::iter::FromIterator;

#[derive(Debug, Serialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct GraphNodesAndEdges<N, E> {
  nodes: Vec<N>,
  edges: Vec<(usize, usize, E)>,
  node_index_to_id: BTreeMap<String, String>,
}

pub type ShapeChoiceMapping = BTreeMap<ShapeId, Vec<ShapeChoice>>;
impl From<&ShapeProjection> for ShapeChoiceMapping {
  fn from(shape_projection: &ShapeProjection) -> Self {
    let choice_queries = ShapeChoiceQueries::from(shape_projection);
    let shape_choices = shape_projection
      .graph
      .node_indices()
      .filter_map(|i| {
        let node = shape_projection
          .graph
          .node_weight(i)
          .expect("node should exist");
        match node {
          Node::Shape(shape_node) => Some(shape_node),
          _ => None,
        }
      })
      .map(|shape_node| {
        let trail = ShapeTrail::new(shape_node.shape_id.clone());
        let choices = choice_queries.trail_choices(&trail).collect();
        (shape_node.shape_id.clone(), choices)
      });
    BTreeMap::from_iter(shape_choices)
  }
}

// duplicated in [projections/spectacle/endpoints.rs]
pub type SerializableGraph = GraphNodesAndEdges<Node, Edge>;
impl From<ShapeProjection> for SerializableGraph {
  fn from(shape_projection: ShapeProjection) -> Self {
    let (graph_nodes, graph_edges) = shape_projection.graph.into_nodes_edges();
    let nodes = graph_nodes.into_iter().map(|x| x.weight).collect();
    let edges = graph_edges
      .into_iter()
      .map(|x| (x.source().index(), x.target().index(), x.weight))
      .collect();
    let value: GraphNodesAndEdges<Node, Edge> = GraphNodesAndEdges {
      nodes,
      edges,
      node_index_to_id: BTreeMap::from_iter(
        shape_projection
          .node_id_to_index
          .into_iter()
          .map(|(k, v)| (v.index().to_string(), k)),
      ),
    };
    value
  }
}

impl ShapeProjection {
  pub fn to_json_string(&self) -> String {
    serde_json::to_string(&self.to_serializable_graph()).expect("graph should serialize")
  }

  pub fn to_serializable_graph(&self) -> SerializableGraph {
    let copy = self.clone();
    copy.into()
  }

  pub fn to_choice_mapping(&self) -> ShapeChoiceMapping {
    self.into()
  }
}
