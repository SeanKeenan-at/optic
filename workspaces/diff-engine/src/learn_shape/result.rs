use crate::commands::shape as shape_commands;
use crate::commands::{ShapeCommand, SpecCommand};
use crate::shapes::JsonTrail;
use crate::state::shape::{ShapeKind, ShapeKindDescriptor};
use crate::BodyDescriptor;
use std::collections::{HashMap, HashSet};

#[derive(Clone, Debug, Default)]
pub struct TrailObservationsResult {
  values_by_trail: HashMap<JsonTrail, TrailValues>,
}

impl TrailObservationsResult {
  pub fn union(&mut self, new_result: TrailObservationsResult) {
    for (json_trail, new_trail_values) in new_result.values_by_trail {
      let existing_trail_values = self
        .values_by_trail
        .entry(json_trail)
        .or_insert_with_key(|json_trail| TrailValues::new(json_trail));

      existing_trail_values.union(new_trail_values);
    }
  }

  pub fn trails(&self) -> impl Iterator<Item = &JsonTrail> {
    self.values_by_trail.keys()
  }

  pub fn values(&self) -> impl Iterator<Item = &TrailValues> {
    self.values_by_trail.values()
  }

  pub fn into_commands<F>(
    mut self,
    mut generate_id: &mut F,
  ) -> (Option<String>, impl Iterator<Item = SpecCommand>)
  where
    F: FnMut() -> String,
  {
    let sorted_trails = {
      let mut trails = self
        .values_by_trail
        .keys()
        .map(|trail| trail.clone())
        .collect::<Vec<_>>();
      trails.sort(); // parents before children
      trails
    };

    let shape_prototypes = sorted_trails
      .into_iter()
      .rev() // leafs first
      .map(|json_trail| {
        let trail_values = self.values_by_trail.remove(&json_trail).unwrap();

        trail_values.into_shape_prototype(&mut generate_id)
      })
      .collect::<Vec<_>>();

    let root_shape_id = shape_prototypes
      .last() // since we created these leafs first, the last must be the root
      .map(|last_shape| last_shape.id.clone());

    let commands =
      shape_prototypes_to_commands(shape_prototypes).map(|command| SpecCommand::from(command));

    (root_shape_id, commands)
  }
}

fn shape_prototypes_to_commands(
  shape_prototypes: impl IntoIterator<Item = ShapePrototype>,
) -> impl Iterator<Item = ShapeCommand> {
  shape_prototypes
    .into_iter()
    .map(|shape_prototype| {
      let [init_commands, describe_commands]: [Option<Vec<ShapeCommand>>; 2] =
        match shape_prototype.prototype_descriptor {
          ShapePrototypeDescriptor::PrimitiveKind { base_shape_kind } => {
            let add_command =
              ShapeCommand::add_shape(shape_prototype.id, base_shape_kind, String::from(""));
            [Some(vec![add_command]), None]
          }
          ShapePrototypeDescriptor::OneOfShape {
            branches,
            parameter_ids,
          } => {
            let mut commands = vec![ShapeCommand::add_shape(
              shape_prototype.id.clone(),
              ShapeKind::OneOfKind,
              String::from(""),
            )];

            let one_off_shape_id = shape_prototype.id.clone();

            let branch_ids = branches
              .iter()
              .zip(parameter_ids)
              .map(|(shape_prototype, parameter_id)| {
                (shape_prototype.id.clone(), parameter_id.clone())
              })
              .collect::<Vec<_>>();

            let branches_commands = shape_prototypes_to_commands(branches)
              .zip(branch_ids)
              .flat_map(|(branch_command, (branch_shape_id, parameter_id))| {
                vec![
                  branch_command,
                  ShapeCommand::add_shape_parameter(
                    parameter_id.clone(),
                    one_off_shape_id.clone(),
                    String::from(""),
                  ),
                  ShapeCommand::set_parameter_shape(
                    one_off_shape_id.clone(),
                    parameter_id.clone(),
                    branch_shape_id.clone(),
                  ),
                ]
              });

            commands.extend(branches_commands);

            [Some(commands), None]
          }
          _ => [None, None],
        };

      [init_commands, describe_commands]
    })
    .fold(
      vec![vec![], vec![]],
      |mut existing_commands, new_commands| {
        let [new_init, new_describe] = new_commands;
        {
          let init_commands = &mut existing_commands[0];
          if let Some(commands) = new_init {
            init_commands.extend(commands);
          }
        }
        {
          let describe_commands = &mut existing_commands[1];

          if let Some(commands) = new_describe {
            describe_commands.extend(commands);
          }
        }
        existing_commands
      },
    )
    .into_iter()
    .flatten()
}

impl From<HashMap<JsonTrail, TrailValues>> for TrailObservationsResult {
  fn from(values_by_trail: HashMap<JsonTrail, TrailValues>) -> Self {
    Self { values_by_trail }
  }
}

pub type FieldSet = HashSet<String>;

#[derive(Clone, Debug)]
pub struct TrailValues {
  pub trail: JsonTrail,
  pub was_string: bool,
  pub was_number: bool,
  pub was_boolean: bool,
  pub was_null: bool,
  pub was_array: bool,
  pub was_object: bool,
  field_set: HashSet<FieldSet>,
}

impl From<JsonTrail> for TrailValues {
  fn from(json_trail: JsonTrail) -> Self {
    TrailValues::new(&json_trail)
  }
}

impl TrailValues {
  pub fn new(json_trail: &JsonTrail) -> Self {
    Self {
      trail: json_trail.clone(),
      was_string: false,
      was_number: false,
      was_boolean: false,
      was_null: false,
      was_array: false,
      was_object: false,
      field_set: Default::default(),
    }
  }

  pub fn union(&mut self, new_values: TrailValues) {
    self.was_string = self.was_string || new_values.was_string;
    self.was_number = self.was_number || new_values.was_number;
    self.was_boolean = self.was_boolean || new_values.was_boolean;
    self.was_null = self.was_null || new_values.was_null;
    self.was_array = self.was_array || new_values.was_array;
    self.was_object = self.was_object || new_values.was_object;
    // TODO: figure out what to do about field sets
  }

  fn into_shape_prototype<F>(self, generate_id: &mut F) -> ShapePrototype
  where
    F: FnMut() -> String,
  {
    let mut descriptors: Vec<_> = vec![
      if self.was_string {
        Some(ShapePrototypeDescriptor::PrimitiveKind {
          base_shape_kind: ShapeKind::StringKind,
        })
      } else {
        None
      },
      if self.was_number {
        Some(ShapePrototypeDescriptor::PrimitiveKind {
          base_shape_kind: ShapeKind::NumberKind,
        })
      } else {
        None
      },
      if self.was_boolean {
        Some(ShapePrototypeDescriptor::PrimitiveKind {
          base_shape_kind: ShapeKind::BooleanKind,
        })
      } else {
        None
      },
      // if self.was_array {
      //   Some(ShapePrototypeDescriptor::ListOfShape {

      //   })
      // } else {
      //   None
      // }
    ]
    .into_iter()
    .flatten()
    .collect();

    let descriptors_count = descriptors.len();
    match descriptors_count {
      0 => ShapePrototype {
        id: generate_id(),
        trail: self.trail,
        prototype_descriptor: ShapePrototypeDescriptor::Unknown,
      },
      1 => ShapePrototype {
        id: generate_id(),
        trail: self.trail,
        prototype_descriptor: descriptors.pop().unwrap(),
      },
      _ => ShapePrototype {
        id: generate_id(),
        trail: self.trail.clone(),
        prototype_descriptor: ShapePrototypeDescriptor::OneOfShape {
          parameter_ids: (0..descriptors.len()).map(|_| generate_id()).collect(),
          branches: descriptors
            .into_iter()
            .map(|descriptor| ShapePrototype {
              id: generate_id(),
              trail: self.trail.clone(),
              prototype_descriptor: descriptor,
            })
            .collect(),
        },
      },
    }
  }
}

#[derive(Debug)]
struct ShapePrototype {
  id: String,
  trail: JsonTrail,
  prototype_descriptor: ShapePrototypeDescriptor,
}

#[derive(Debug)]
enum ShapePrototypeDescriptor {
  OptionalShape {
    shape: Box<ShapePrototype>,
  },
  NullableShape {
    shape: Box<ShapePrototype>,
  },
  OneOfShape {
    branches: Vec<ShapePrototype>,
    parameter_ids: Vec<String>,
  },
  ObjectWithFields {
    fields: Vec<ShapePrototype>,
  },
  ListOfShape {
    shape: Box<ShapePrototype>,
  },
  FieldWithShape {
    key: String,
    shape: Box<ShapePrototype>,
  },
  PrimitiveKind {
    base_shape_kind: ShapeKind,
  },
  Unknown,
}

#[cfg(test)]
mod test {
  use super::*;
  use crate::learn_shape::observe_body_trails;
  use crate::projections::SpecProjection;
  use crate::state::body::BodyDescriptor;
  use cqrs_core::Aggregate;
  use insta::assert_debug_snapshot;
  use serde_json::json;

  #[test]
  fn trail_observations_can_generate_commands_for_primitive_bodies() {
    let string_body = BodyDescriptor::from(json!("a string body"));
    let number_body = BodyDescriptor::from(json!(48));
    let boolean_body = BodyDescriptor::from(json!(true));

    let string_observations = observe_body_trails(string_body);
    let number_observations = observe_body_trails(number_body);
    let boolean_observations = observe_body_trails(boolean_body);

    let mut counter = 0;
    let mut test_id = || {
      let id = format!("test-id-{}", counter);
      counter += 1;
      id
    };
    let spec_projection = SpecProjection::default();

    let string_results = collect_commands(string_observations.into_commands(&mut test_id));
    assert!(string_results.0.is_some());
    assert_eq!(string_results.1.len(), 1);
    spec_projection
      .execute((&string_results.1[0]).clone())
      .expect("generated command should be valid");
    assert_debug_snapshot!(
      "trail_observations_can_generate_commands_for_primitive_bodies__string_results",
      &string_results
    );

    let number_results = collect_commands(number_observations.into_commands(&mut test_id));
    assert!(number_results.0.is_some());
    assert_eq!(number_results.1.len(), 1);
    spec_projection
      .execute((&number_results.1[0]).clone())
      .expect("generated command should be valid");
    assert_debug_snapshot!(
      "trail_observations_can_generate_commands_for_primitive_bodies__number_results",
      number_results
    );

    let boolean_results = collect_commands(boolean_observations.into_commands(&mut test_id));
    assert!(boolean_results.0.is_some());
    assert_eq!(boolean_results.1.len(), 1);
    spec_projection
      .execute((&boolean_results.1[0]).clone())
      .expect("generated command should be valid");
    assert_debug_snapshot!(
      "trail_observations_can_generate_commands_for_primitive_bodies__boolean_results",
      boolean_results
    );
  }

  #[test]
  fn trail_observations_can_generate_commands_for_one_off_of_primitives() {
    let string_body = BodyDescriptor::from(json!("a string body"));
    let number_body = BodyDescriptor::from(json!(48));
    let boolean_body = BodyDescriptor::from(json!(true));

    let mut observations = TrailObservationsResult::default();
    observations.union(observe_body_trails(string_body));
    observations.union(observe_body_trails(number_body));
    observations.union(observe_body_trails(boolean_body));

    let mut counter = 0;
    let mut test_id = || {
      let id = format!("test-id-{}", counter);
      counter += 1;
      id
    };
    let mut spec_projection = SpecProjection::default();

    let observation_results = collect_commands(observations.into_commands(&mut test_id));
    assert!(observation_results.0.is_some());
    assert_debug_snapshot!(
      "trail_observations_can_generate_commands_for_one_off_of_primitives__observation_results",
      observation_results
    );

    for command in observation_results.1 {
      let events = spec_projection
        .execute(command)
        .expect("generated commands must be valid");

      for event in events {
        spec_projection.apply(event)
      }
    }
  }

  fn collect_commands(
    (root_shape_id, commands): (Option<String>, impl Iterator<Item = SpecCommand>),
  ) -> (Option<String>, Vec<SpecCommand>) {
    (root_shape_id, commands.collect::<Vec<_>>())
  }
}
