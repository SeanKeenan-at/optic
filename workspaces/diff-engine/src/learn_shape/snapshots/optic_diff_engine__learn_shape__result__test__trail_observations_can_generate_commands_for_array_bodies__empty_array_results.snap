---
source: workspaces/diff-engine/src/learn_shape/result.rs
expression: "&empty_array_results"
---
(
    Some(
        "test-id-2",
    ),
    [
        ShapeCommand(
            AddShape(
                AddShape {
                    shape_id: "test-id-2",
                    base_shape_id: "$list",
                    name: "",
                },
            ),
        ),
    ],
)
